packages/grafana-ui/src/components/NodeGraph/utils.ts

```diff
@@ -0,0 +1,67 @@
+import { DataFrame, FieldCache } from '@grafana/data';
+
+type Line = { x1: number; y1: number; x2: number; y2: number };
+
+/**
+ * Makes line shorter while keeping the middle in he same place.
+ */
+export function shortenLine(line: Line, length: number): Line {
+  const vx = line.x2 - line.x1;
+  const vy = line.y2 - line.y1;
+  const mag = Math.sqrt(vx * vx + vy * vy);
+  const ratio = Math.max((mag - length) / mag, 0);
+  const vx2 = vx * ratio;
+  const vy2 = vy * ratio;
+  const xDiff = vx - vx2;
+  const yDiff = vy - vy2;
+  const newx1 = line.x1 + xDiff / 2;
+  const newy1 = line.y1 + yDiff / 2;
+  return {
+    x1: newx1,
+    y1: newy1,
+    x2: newx1 + vx2,
+    y2: newy1 + vy2,
+  };
+}
+
+export function getNodeFields(nodes: DataFrame) {
```

Comment:
@torkelo @ryantxu @dprokop this is the main logic how the dataFrame is parsed and fields assigned to particular placeholders. I use constant label values to mark any column and then it's used later on. This means it is possible to send columns with what ever names/values so it would make sense for example in a table view, for example to get Node with highest percentage of errors and then just label some of them to be used for service map. It's also possible to `NodeGraphDetailField` label to add any fields into the context menu header. This should be flexible enough for jaeger/zipkin/tempo later on.

Main downside is that data source will have to add dependency for this (maybe the constants should be in grafana/data but considering this is UI concern I am not sure).

Another thing is that maybe formatting is bit lacking, not sure how to easily add thing like prefix (in x-ray there is `avg. 150ms` as one of the stat inside the node. With this we can just add string type column and have that `avg.` prefix there but seems a bit awkward)

Reply:
Here is a relevant code on the x-ray side of things https://github.com/grafana/x-ray-datasource/pull/60/files#diff-58036290d63f45dcebfa80ddeeb5877187a613f7cdb28e9a5482324f632c14adR225